lib/connectors/memory.js

```diff
@@ -250,6 +250,43 @@ Memory.prototype.updateOrCreate = function(model, data, options, callback) {
     }
   });
 };
+Memory.prototype.patchOrCreateWithWhere =
+  Memory.prototype.upsertWithWhere = function(model, where, data, options, callback) {
+    var self = this;
+    var element = Object.keys(where)[0];
+    var filter = { where: where };
+    if (!data.hasOwnProperty(element)) {
+      data[element] = where[element];
+    }
+    var primaryKey = this.idName(model);
+    self.all(model, filter, options, function(err, instances) {
```

Comment:
`self.all` is non-atomic. There is no point in implementing `upsertWithWhere` in memory connector that is not atomic, that would be just the same as we already have in DAO. What we need instead, is an atomic version. Check how `findOrCreate` is implemented using `__findAllSkippingIncludes`.